compiler_opt/tools/regalloc_trace/extract_functions_lib.py

```diff
@@ -0,0 +1,163 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""A library that contains utilities for extracting functions."""
+
+import subprocess
+import os
+import shutil
+import json
+import concurrent.futures
+from collections.abc import Sequence
+
+
+def _get_functions_in_file(bc_file_path: str, llvm_nm_path: str) -> list[str]:
+  """Gets all function names defined in a file.
+
+  Args:
+    bc_file_path: The path to the bitcode file to find functions in.
+    llvm_nm_path: The path to the llvm-nm binary that is used to extract all of
+      the symbol names within the function.
+
+  Returns:
+    A list of strings representing the functions in the file.
+  """
+  command_vector = [
+      llvm_nm_path,
+      "--defined-only",
+      "--format=posix",
+      bc_file_path,
+  ]
+  result = subprocess.run(command_vector, capture_output=True, check=True)
+
+  functions_list = []
+  for symbol in result.stdout.decode("utf-8").split("\n")[:-1]:
+    symbol_parts = symbol.split(" ")
+    if (symbol_parts[1] == "t" or symbol_parts[1] == "T" or
+        symbol_parts[1] == "w" or symbol_parts[1] == "W"):
+      functions_list.append(symbol_parts[0])
+
+  return functions_list
+
+
+def _extract_function_from_file(
+    bc_file_path: str,
+    output_file_path: str,
+    function_name: str,
+    llvm_extract_path: str,
+    opt_path: str,
+) -> None:
+  """Extracts a function from a file.
+
+  Args:
+    bc_file_path: The path to the bitcode file to extract the function from.
+    output_folder: The folder to dump the extracted function into.
+    function_name: The (mangled) name of the function to extract.
+    llvm_extract_path: The path to the llvm-extract binary to use to extract
+      the function.
+    opt_path: The path to the opt binary to use to strip debug information.
+  """
+  command_vector = [
+      llvm_extract_path,
+      "-func",
+      function_name,
+      "-o",
+      output_file_path + ".fat",
+      bc_file_path,
+  ]
+  subprocess.run(command_vector, capture_output=True, check=True)
+
+  opt_command_vector = [
+      opt_path,
+      "-strip-debug",
+      "-o",
+      output_file_path,
+      output_file_path + ".fat",
+  ]
+  subprocess.run(opt_command_vector, capture_output=True, check=True)
+  os.remove(output_file_path + ".fat")
+
+  orig_cmd_file_path = bc_file_path[:-3] + ".cmd"
```

Comment:
More readable & generic wrt extension size: `os.path.splitext(bc_file_path)[0] + ".cmd"`

here and below